### PR TITLE
Add support for inherited sockets

### PR DESCRIPTION
**Pull request by Cu3PO42**
This PR implements support for using inherited sockets instead of opening a new socket on a specified port. The primary motivation for this feature is [socket activation](http://0pointer.de/blog/projects/socket-activation.html). It lets the Robocode server be started on demand as soon as a bot, controller or observer conects to it.

I also intend to use this feature to make orchestration of the server easier without dealing with port conflicts for my battle runner.

Finally, this also enables (indirect) IPv6 support by opening a v6 socket and passing it.

----------------

Currently, this relies on an unreleased version of Java WebSocket, so I have marked this as a draft. I'll update the PR once v1.6 of the library is released.

### CODE REVIEW

**Comment by flemming-n-larsen** [comment]
@Cu3PO42 Really nice. I will try this out to see how it works. And yes, we might need to wait with releasing it until the unreleased version of Java WebSocket becomes a real release, as it might have some unwanted side-effects. Let's see. 🙂

**Comment by Cu3PO42** [comment]
Thanks for taking a look! The easiest way to test this is to run `systemd-socket-activate -l 7654 --inetd java -jar path/to/server.jar -p inherit` on a Linux  system with systemd. Passing sockets is supported on other OSes as well, but you would probably need to implement your own helper to test it.

**Comment by flemming-n-larsen** [comment]
@Cu3PO42 I have had a first look at the changes you made, and it is looking good to me.
And I am very pleased that you also updated the description of the port parameter etc. so it fits the new feature. 👍 

Since I am working mostly on a Windows 11 machine, I will setup Ubuntu on my system (again) so I can try out how it works. 🙂 I am pretty sure it will work just fine.

While running the code locally, I found another issue with the server that I will try to fix. I hope it will not impact this PR.

**Comment by Cu3PO42** [comment]
Thank you! This feature should work just fine inside WSL2, so you might get away with simply installing that. If I recall correctly, running `wsl` on the command line should prompt you to install any necessary components.

**Comment by SirStone** [comment]
Just a question about this implementation. Would be possible that socket activation could allow a bot to be reloaded without loosing the connection? 
I was thinking that would be really nice to be able to restart the bot, after changes to it's code, without requiring to stop the bot->stop current match->add the bot just connected again->start the match at every little change

**Comment by Cu3PO42** [comment]
> Would be possible that socket activation could allow a bot to be reloaded without loosing the connection?

Not really, no. You *could* use the same mechanism to pass a socket from one generation of your bot to the new version after you make some changes, but an open socket isn't all that useful. You would need an open WebSocket connection, which has a bunch of associated state. It is possible to also preserve that when `execve`-ing a new binary, but it would be very complicated.

A much more realistic approach for your scenario would be to write a thin proxy for the connection between bot and server. That should be possible with relatively little effort and honestly sounds like a good idea. This would be entirely unrelated to the changes introduced here, though.

**Comment by SirStone** [comment]
Thank you for your reply. I may consider creating a proxy later if it
proves to be more efficient than my current approach, which involves
several intermediate steps to observe changes in my bots during the game.

If others show interest, they might encourage me to do it or even take on
the task themselves.

Thanks, and sorry for the digression from the main topic

Il giorno dom 1 dic 2024 alle ore 16:56 Tobias Zimmermann <
***@***.***> ha scritto:

**Comment by marci4** [comment]
> @Cu3PO42 Really nice. I will try this out to see how it works. And yes, we might need to wait with releasing it until the unreleased version of Java WebSocket becomes a real release, as it might have some unwanted side-effects. Let's see. 🙂

I released 1.6.0 yesterday :)

**Comment by flemming-n-larsen** [comment]
@marci4:

> I released 1.6.0 yesterday :)

Nice! Thank you for the update/notification! 😊👍

@Cu3PO42
I will update the dependency. Test it out, and merge the pull request later today. 😊

**Comment by Cu3PO42** [comment]
Thank you both! I'm excited to see this merged :-)

File: server/src/main/kotlin/dev/robocode/tankroyale/server/connection/MultiServerWebSocketObserver.kt [hunk 1]
```diff
@@ -2,12 +2,20 @@ package dev.robocode.tankroyale.server.dev.robocode.tankroyale.server.connection
...
 import java.net.InetSocketAddress
+import kotlin.system.exitProcess
 
 class MultiServerWebSocketObserver(observer: IClientWebSocketObserver) {
 
+    private val log = LoggerFactory.getLogger(this::class.java)
+
     private val webSocketServer = if (Server.useInheritedChannel) {
+        if (System.inheritedChannel() == null) {
+            log.error("'${Server.INHERIT}' (socket activation) is not supported on this system.")
```
Comment (Cu3PO42) [inline server/src/main/kotlin/dev/robocode/tankroyale/server/connection/MultiServerWebSocketObserver.kt:16]: Just some small feedback on this error message: `System.inheritedChannel()` will be null if starting the server without having passed a socket as fd 3. This doesn't mean this mode of operation is inherently unsupported on this system, just that the user didn't start the application correctly.

I would recommend messaging along the following lines: `'${Server.INHERIT}' can only be used when a socket has been passed as stdin; if you did pass a socket, this feature may be unsupported on this system`

For additional safety, you might want to check that `System.inheritedChannel() instanceof ServerSocketChannel`. If this check fails, but it is `!= null`, then the user did definitely invoke the application incorrectly.

**Comment by flemming-n-larsen** [comment]
@Cu3PO42

So now I want to test this feature? 🙂
I guess it should be easy on Linux and macOS, but perhaps not Windows?

**Comment by Cu3PO42** [comment]
Hi! I am so sorry, I have been swallowed up by life and work for the last couple of days. I did some research into socket inheritance on Windows and it is possible, but much more complicated than on Unix systems. It also requires a different, much more involved form of cooperation and does not seem to work with Java at all (without ugly JNI hacks at least).

You are on the right track with your current configuration. There is a somewhat easier way to test this by simply running `systemd-socket-activate -l 7654 --inetd java -jar path/to/server.jar -p inherit`. This does not require setting up any of the files that you did, but is not persistent.

As for the issue you are currently experiencing, it is because Java follows an older convention for passing sockets than systemd defaults to. You need the following in your service file:

```
[Unit]
Description=Robocode Tank Royale Service

[Service]
ExecStart=/usr/bin/java -jar /opt/robocode/robocode-tankroyale-server-0.28.0.jar --port=inherit
StandardInput=socket
```

**Comment by flemming-n-larsen** [comment]
@Cu3PO42 
Your work is now included in release [0.28.0](https://github.com/robocode-dev/tank-royale/releases/tag/v0.28.0). Thank your for bringing this cool feature to Robocode. 💪

@marci4
Thank you for for providing and maintaining the excellent [Java WebSocket](https://tootallnate.github.io/Java-WebSocket/) library, making WebSockets accessible for applications like Robocode without the need of running a "real" server. 🙏

**Pull request merged**